src/sanitation.js

```diff
@@ -239,15 +254,23 @@ export function isValidAttr(tagName, attrName, attrValue, opt_purify = false) {
     return false;
   }
 
+  let attrBlacklist;
+  let attrNameBlacklist;
+  if (isAmp4Email_(opt_doc)) {
+    attrBlacklist = STRICT_BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES[tagName];
+    attrNameBlacklist = AMP4EMAIL_BLACKLISTED_TAG_SPECIFIC_ATTRS[tagName];
+  } else {
+    attrBlacklist = BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES[tagName];
+    attrNameBlacklist = BLACKLISTED_TAG_SPECIFIC_ATTRS[tagName];
+  }
```

Comment:
Let's merge the objects here. This avoids the class of bugs where we forget to update one but not the other.

Reply:
The `else` case here is unnecessary. I think this can be simplified:

```js
const attrNameBlacklist = BLACKLISTED_TAG_SPECIFIC_ATTRS[tagName];
if (isAmp4Email(doc)) {
  Object.assign(attrNameBlacklist, AMP4EMAIL_BLACKLISTED_TAG_SPECIFIC_ATTRS[tagName]);
}
if (attrNameBlacklist && attrNameBlacklist.indexOf(attrName) != -1) {
  return false;
}
```

Ditto for `attrBlacklist`.

Reply:
done

Reply:
Actually, this might need to be:
```js
 Object.assign(dict(), attrNameBlacklist, AMP4EMAIL...);
```
Setting `attrNameBlacklist` as the target might mutate `BLACKLISTED_TAG_SPECIFIC_ATTRS`.

We should consider freezing these const objects.

Reply:
Right, I actually noticed and made that switch in the code change. It is now

    Object.assign(AMP4EMAIL..., attrNameBlacklist);

Which does mutate the AMP4Email blacklist. Given that the only thing reading these values is this method, I'll just inline the AMP4EMAIL blacklist data here which would be safer.

Reply:
can you explain? The change I made is the following:

      attrNameBlacklist = (/** @type {!JsonObject} */ (Object.assign(
        {'form': ['name'], 'amp-anim': ['controls']}, attrNameBlacklist)))
        [tagName];

I just add the existing attribute name blacklist (attrNameBlacklist) rules to the AMP4Email specific rules, creating a new blacklist and leaving attrNameBlacklist intact.

Reply:
I meant this:

```js
Object.assign(AMP4EMAIL..., attrNameBlacklist); // BAD, email rules don't take precedence.
Object.assign(attrNameBlacklist, AMP4EMAIL); // OK
```

Overall I was trying to make this code path more readable. Suggested:

```js
  // Remove blacklisted attributes from specific tags e.g. input[formaction].
  const attrBlacklist = map(BLACKLISTED_ATTRS[tagName]);
  if (isAmp4Email(doc)) {
    Object.assign(attrBlacklist, EMAIL_BLACKLISTED_ATTRS[tagName]);
  }
  if (attrBlacklist && attrBlacklist.indexOf(attrName) != -1) {
    return false;
  }

  // Remove blacklisted values for specific attributes for specific tags e.g. input[type=image].
  const attrValueBlacklist = map(BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES[tagName]);
  if (isAmp4Email(doc)) {
    Object.assign(attrValueBlacklist, EMAIL_BLACKLISTED_TAG_SPECIFIC_ATTR_VALUES[tagName]);
  }
  if (attrValueBlacklist) {
    const blacklistedValuesRegex = attrBlacklist[attrName];
    if (blacklistedValuesRegex &&
        attrValue.search(blacklistedValuesRegex) != -1) {
      return false;
    }
  }
```

Almost everything can stay the same, except we're adding the `if (isAmp4Email(doc)) {` conditional twice.

Reply:
Oh ok I see what you're saying. So rather than assigning the values from one map to the other, I'll just add the blacklisted attributes to that existing blacklisted list, as there is no overlapping of data. When the time comes when AMP4Email allows an attribute that is blacklisted in the AMP format then we maintain two separate maps and do Object.assign. 

Note that I didn't word for word follow your example due some to some errors but followed the two checks method. Checking for the tag attribute value blacklist first then the tag attribute value.

Hopefully it's more readable now.